tests/c_api_tests/test_cpp_api_model.cpp

```diff
@@ -0,0 +1,269 @@
+                                               .sk = 1000.0,
+                                               .rx_ratio = 0.0,
+                                               .z01_ratio = 1.0};
+    power_grid_model::SymLoadGenInput load_input{.id = 2,
+                                                 .node = 0,
+                                                 .status = 1,
+                                                 .type = power_grid_model::LoadGenType::const_i,
+                                                 .p_specified = 0.0,
+                                                 .q_specified = 500.0};
+
+    // create one buffer and set attr, leave angle to nan as default zero, leave z01 ratio to nan
+    Buffer source_buffer{PGM_def_input_source, 1};
+    source_buffer.set_nan(0);
+    source_buffer.set_value(PGM_def_input_source_id, &source_input.id, 0, -1);
+    source_buffer.set_value(PGM_def_input_source_node, &source_input.node, 0, sizeof(ID));
+    source_buffer.set_value(PGM_def_input_source_status, &source_input.status, 0, -1);
+    source_buffer.set_value(PGM_def_input_source_u_ref, &source_input.u_ref, 0, -1);
+    source_buffer.set_value(PGM_def_input_source_sk, &source_input.sk, 0, -1);
+    source_buffer.set_value(PGM_def_input_source_rx_ratio, &source_input.rx_ratio, 0, -1);
+
+    // add buffer
+    input_dataset.add_buffer("node", 1, 1, nullptr, &node_input);
+    input_dataset.add_buffer("sym_load", 1, 1, nullptr, &load_input);
+    input_dataset.add_buffer("source", 1, 1, nullptr, source_buffer);
+
+    // output data
+    std::array<power_grid_model::NodeOutput<power_grid_model::symmetric_t>, 2> sym_node_outputs{};
+    power_grid_model::NodeOutput<power_grid_model::symmetric_t>& node_result_0 = sym_node_outputs[0];
+    power_grid_model::NodeOutput<power_grid_model::symmetric_t>& node_result_1 = sym_node_outputs[1];
+    DatasetMutable single_output_dataset{"sym_output", 0, 1};
+    single_output_dataset.add_buffer("node", 1, 1, nullptr, sym_node_outputs.data());
+    DatasetMutable batch_output_dataset{"sym_output", 1, 2};
+    batch_output_dataset.add_buffer("node", 1, 2, nullptr, sym_node_outputs.data());
+
+    // update data
+    power_grid_model::SourceUpdate source_update{
+        .id = 1, .status = power_grid_model::na_IntS, .u_ref = 0.5, .u_ref_angle = power_grid_model::nan};
+    std::array<Idx, 3> source_update_indptr{0, 1, 1};
+    std::array<power_grid_model::SymLoadGenUpdate, 2> load_updates{};
+    // set nan twice with offset
+    Buffer::set_nan(PGM_def_update_sym_load, load_updates.data(), 0, 1);
+    Buffer::set_nan(PGM_def_update_sym_load, load_updates.data(), 1, 1);
+    // set value
+    load_updates[0].id = 2;
+    load_updates[0].q_specified = 100.0;
+    load_updates[1].id = 2;
+    load_updates[1].q_specified = 300.0;
+    // dataset
+    DatasetConst single_update_dataset{"update", 0, 1};
+    single_update_dataset.add_buffer("source", 1, 1, nullptr, &source_update);
```

Comment:
here and similarly in other places: can you use the dataset definitions like

```suggestion
    single_update_dataset.add_buffer(PGM_dataset_definitions_source, 1, 1, nullptr, &source_update);
```